ffi/framebuffer.lua

```diff
@@ -138,13 +138,21 @@ function fb:setViewport(viewport)
     or viewport.w < 0 or viewport.x+viewport.w > self.screen_size.w
     or viewport.h < 0 or viewport.y+viewport.h > self.screen_size.h
     then
-        error("bad viewport")
+        error("fb:setViewport() bad viewport")
     end
-    self.full_bb = self.bb
-    self.bb = self.bb:viewport(
+    -- we might be switching the viewport on the fly, so don't overwrite self.full_bb
+    if not self.full_bb then
+        self.full_bb = self.bb
+    end
+    self.debug("fb:setViewport() setting viewport to",
+               viewport.x, viewport.y,
+               viewport.w, viewport.h)
+    self.bb = self.full_bb:viewport(
         viewport.x, viewport.y,
         viewport.w, viewport.h)
     self.viewport = viewport
+    self.full_bb:fill(Blitbuffer.COLOR_WHITE)
```

Comment:
@poire-z Okay, it all looks good this way. Do you think these extra two lines are a self-evident addition to setViewport or some kind of overreach?

Reply:
From nowhere, I'd ask: why do you need them? You're setting a viewport on an existing full_bb, why would you need to erase it and repaint it? But if you did it, it obviously helps in some way, no?

Reply:
You didn't see my screenshot? The existing full_bb already has content on it. It'll just remain there in perpetuity.

The only question is where these lines should be called from. Not if.

Reply:
Oh, ok, I probably don't get what viewport do (thought you had a full_bb, and viewport was the part of it that was displayed, so stuff outside would not matter).
So, you're more knowing than me to answer this question :)

Reply:
Traditionally what you say is essentially correct.

1. We have a device, such as the H2O.
2. We set the viewport on init combined with a white fill. This is enough. The white will stay white.

But now we add:

3. We dynamically adjust the viewport on Android to adjust our display for the statusbar/toolbar. (I'd prefer to separate them out rather than just a generic "fullscreen".)

Whether this is 100% for sure the best way to do it I'm not entirely sure, but it's the easiest to implement for now and all roadblocks along the way are relevant with or without viewport (e.g., resizable SDL window).

Only the ReaderUI is currently capable of dynamic adjustment due to portrait/landscape. It'll need to be added to the FileManager. But I'll ignore that problem for now. Especially since there are some issues to be worked out on the Android side of things that I want to be as little involved in as possible.